pkg/controller/direct/compute/firewallpolicyrule/firewallpolicyrule_controller.go

```diff
@@ -0,0 +1,290 @@
+	// TODO: Support URLs
+	return nil, nil
+}
+
+func (a *firewallPolicyRuleAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting ComputeFirewallPolicyRule", "priority", a.priority)
+
+	firewallPolicyRule, err := a.get(ctx)
+	if err != nil {
+		// When a certain rule does not exist, the error has code 400(invalid) instead of 404(not found)
+		// example error message:
+		// "Invalid value for field 'priority': '9000'. The firewall policy does not contain a rule at priority 9000.",
+		if direct.IsInvalidValue(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+	a.actual = firewallPolicyRule
+	return true, nil
+}
+
+func (a *firewallPolicyRuleAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	var err error
+
+	err = resolveDependencies(ctx, a.reader, a.desired)
+	if err != nil {
+		return err
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating ComputeFirewallPolicyRule", "priority", a.priority)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+
+	firewallPolicyRule := ComputeFirewallPolicyRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &computepb.AddRuleFirewallPolicyRequest{
+		FirewallPolicyRuleResource: firewallPolicyRule,
+		FirewallPolicy:             a.firewallPolicy,
+	}
+	op, err := a.firewallPoliciesClient.AddRule(ctx, req)
+
+	if err != nil {
+		return fmt.Errorf("creating ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+	if !op.Done() {
+		err = op.Wait(ctx)
+		if err != nil {
+			return fmt.Errorf("waiting ComputeFirewallPolicyRule %d create failed: %w", a.priority, err)
+		}
+	}
+	log.V(2).Info("successfully created ComputeFirewallPolicyRule", "priority", a.priority)
+
+	// Get the created resource
+	created := &computepb.FirewallPolicyRule{}
+	created, err = a.get(ctx)
+	if err != nil {
+		return fmt.Errorf("getting ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+
+	status := &krm.ComputeFirewallPolicyRuleStatus{
+		RuleTupleCount: direct.PtrTo(int64(*created.RuleTupleCount)),
+		Kind:           direct.PtrTo("compute#firewallPolicyRule"),
+	}
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *firewallPolicyRuleAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	var err error
+
+	err = resolveDependencies(ctx, a.reader, a.desired)
+	if err != nil {
+		return err
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating ComputeFirewallPolicyRule", "priority", a.priority)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	firewallPolicyRule := ComputeFirewallPolicyRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	firewallPolicyRule.Priority = nil
```

Comment:
Shall we put this inside the mapper function? If there's a reason to not to, could you add a comment?

Reply:
Added a comment. For patch request, the priority field is not a part of the request body, but a query parameter.